plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange/ASTPreProcess.hs

```diff
@@ -61,19 +63,18 @@ mergeImports node = pure $ node { nodeChildren = children }
         . nodeChildren $ node
 
     merge :: [HieAST a] -> Maybe (HieAST a)
-    merge []  = Nothing
-    merge [x] = Just x
-    merge xs  = createVirtualNode xs
+    merge []     = Nothing
+    merge [x]    = Just x
+    merge (x:xs) = Just $ createVirtualNode (x NonEmpty.:| xs)
 
 nodeIsImport :: HieAST a -> Bool
 nodeIsImport = isAnnotationInAstNode ("ImportDecl", "ImportDecl")
 
-createVirtualNode :: [HieAST a] -> Maybe (HieAST a)
-createVirtualNode [] = Nothing
-createVirtualNode children = Just $ mkAstNode (NodeInfo mempty mempty mempty) span' children
+createVirtualNode :: NonEmpty (HieAST a) -> HieAST a
+createVirtualNode children = mkAstNode (NodeInfo mempty mempty mempty) span' (NonEmpty.toList children)
   where
     span' :: RealSrcSpan
-    span' = foldl1' combineRealSrcSpans . fmap nodeSpan $ children
+    span' = foldl1 combineRealSrcSpans . fmap nodeSpan $ children
```

Comment:
I meant this one! https://hackage.haskell.org/package/semigroupoids-5.3.7/docs/Data-Semigroup-Foldable.html#v:fold1

The reason I'm being picky is that at some point I'm hoping we can just ban e.g. `foldl1` from `Foldable`. It is safe here, but it's easier if we can rely on people using the safe one from `Foldable1` when it's safe.

Reply:
Oh, thanks, I didn't know this function. Changed it to `foldlM1` in `Foldable1` though, as `fold1` requires a semigroup instance, and using the existing `Identity` monad is easier.